### DISCUSSION TITLE
Sequence Parallelism,  memory usage question

### DISCUSSION THREAD

**Discussion by EthanChen1234** [hpcaitech/ColossalAI#3803] [Community | General]
the paper, "Sequence Parallelism: Long Sequence Training from System Perspective"

![image](https://github.com/hpcaitech/ColossalAI/assets/45474996/07dba89d-48e4-430d-b022-6da98d4ad673)

for the tensor parallelism, the activations and trainable weight consume memory.
the activations, contain 4BLH/N and BLH. the trainable weight should be 4H^2/N + 4H^2/N = 8H^2/N.

I'm confused, the  32H^2/N memory use.

**Comment by EthanChen1234** [comment]
I'm still waiting.

**Reply by FrankLeeeee** [reply]
Hi, it 32H^2/N refers to the space occupied by M2 which is the weight of the linear layer. It is 32 because we count in the gradients, momentum, variance assuming using Adam optimizer instead of the weight alone.

**Reply by 0oshowero0** [reply]
Hi, I'm also confused about this. In my calcualtion (according to ZeRO), the memory usage for M2 should be:
model parameters: $\frac{4H^2}{N}$
gradients: $\frac{4H^2}{N}$
optimizer states: $\frac{4H^2}{N} * 3$ (parameters, momentum, variance).

So it should be $\frac{20H^2}{N}$ for the 1st/2st linear and  $\frac{40H^2}{N}$ together.
